packages/documentsearch/src/tokens.ts

```diff
@@ -13,26 +14,300 @@ export const ISearchProviderRegistry = new Token<ISearchProviderRegistry>(
+   */
+  filters: IFiltersType;
+
+  /**
+   * Is the filters view open?
+   */
+  filtersOpen: boolean;
+}
+
+/**
+ * Base search match interface
+ */
+export interface ISearchMatch {
+  /**
+   * Text of the exact match itself
+   */
+  readonly text: string;
+
+  /**
+   * Start location of the match (in a text, this is the column)
+   */
+  position: number;
+}
+
+/**
+ * HTML search match interface
+ */
+export interface IHTMLSearchMatch extends ISearchMatch {
+  /**
+   * Node containing the match
+   */
+  readonly node: Text;
+}
+
+/**
+ * Interface for search provider factory
+ */
+export interface ISearchProviderFactory<T extends Widget = Widget> {
+  /**
+   * Instantiate a search provider for the widget.
+   *
+   * #### Notes
+   * The widget provided is always checked using `isApplicable` before calling
+   * this factory.
+   *
+   * @param widget The widget to search on
+   * @param translator [optional] The translator object
+   *
+   * @returns The search provider on the widget
+   */
+  readonly createNew: (widget: T, translator?: ITranslator) => ISearchProvider;
+
+  /**
+   * Report whether or not this provider has the ability to search on the
+   * given widget. The function is a type guard, meaning that it returns
+   * a boolean, but has a type predicate (`x is T`) for its return signature.
+   *
+   * @param domain Widget to test
+   */
+  readonly isApplicable: (domain: Widget) => domain is T;
+}
+
+/**
+ * Search provider registry interface
+ */
 export interface ISearchProviderRegistry {
   /**
    * Add a provider to the registry.
    *
    * @param key - The provider key.
    * @returns A disposable delegate that, when disposed, deregisters the given search provider
    */
-  register(key: string, provider: ISearchProviderConstructor<any>): IDisposable;
+  add(key: string, provider: ISearchProviderFactory<Widget>): IDisposable;
 
   /**
    * Returns a matching provider for the widget.
    *
    * @param widget - The widget to search over.
    * @returns the search provider, or undefined if none exists.
    */
-  getProviderForWidget(widget: any): ISearchProvider<any> | undefined;
+  getProvider(widget: Widget): ISearchProvider | undefined;
 
   /**
    * Signal that emits when a new search provider has been registered
    * or removed.
    */
   changed: ISignal<ISearchProviderRegistry, void>;
 }
+
+/**
+ * Base search provider interface
+ *
+ * #### Notes
+ * It is implemented by subprovider like searching on a single cell.
+ */
+export interface IBaseSearchProvider extends IDisposable {
+  /**
+   * Start a search
+   *
+   * @param query Regular expression to test for
+   * @param filters Filters to apply when searching
+   */
+  startQuery(query: RegExp, filters: IFiltersType): Promise<void>;
+
+  /**
+   * Stop a search and clear any internal state of the provider
+   */
+  endQuery(): Promise<void>;
+
+  /**
+   * Clear currently highlighted match.
+   */
+  clearHighlight(): void;
```

Comment:
Will this always be trivial to implement synchronously?

Since the other functions return promises, it's low-cost to make this one return a promise if it'll ever be useful. But if it's always safe to assume this can be done synchronously then fair enough :+1: